sbi/inference/posteriors/vi_posterior.py

```diff
@@ -1,69 +1,481 @@
+            show_progress_bar: If any progress report should be displayed.
+            quality_controll_metric: Which metric to use for evaluate the quality.
+            kwargs: Hyperparameters
+                retain_graph: Boolean which decides weather to retain the computation
+                    graph. This may be required for some 'exotic' user-specified q's.
+        Returns:
+            NeuralPosterior: The VIPosterior (can be used to chain calls).
+        """
+
+        # Init q and the optimizer if necessary
+        if retrain_from_scratch:
+            self.set_q(self._q_arg, self._q_kwargs)
+            self._optimizer = self._optimizer_base(
+                self.potential_fn,
+                self.q,
+                lr=learning_rate,
+                clip_value=clip_value,
+                gamma=gamma,
+                n_particles=n_particles,
+                **kwargs,
+            )
+
+        if (
+            reset_optimizer
+            or self._optimizer is None
+            or not isinstance(self._optimizer, self._optimizer_base)
+        ):
+            self._optimizer = self._optimizer_base(
+                self.potential_fn,
+                self.q,
+                lr=learning_rate,
+                clip_value=clip_value,
+                gamma=gamma,
+                n_particles=n_particles,
+                **kwargs,
+            )
+
+        # Check context
+        x = atleast_2d_float32_tensor(self._x_else_default_x(x)).to(self._device)
+        already_trained = self._trained_on is not None and (x == self._trained_on).all()
+
+        # Optimize
+        self._optimizer.update({**locals(), **kwargs})
+        optimizer = self._optimizer
+        optimizer.to(self._device)
+        optimizer.reset_loss_stats()
+
+        if show_progress_bar:
+            iters = tqdm(range(max_num_iters))
+        else:
+            iters = range(max_num_iters)
+
+        # Warmup before training
+        if reset_optimizer or (not optimizer.warm_up_was_done and not already_trained):
+            if show_progress_bar:
+                iters.set_description("Warmup phase, this takes some seconds...")
+            optimizer.warm_up(warm_up_rounds)
+
+        for i in iters:
+            optimizer.step(x)
+            mean_loss, std_loss = optimizer.get_loss_stats()
+            # Update progress bar
+            if show_progress_bar:
+                iters.set_description(
+                    f"Loss: {np.round(mean_loss, 2)} Std: {np.round(std_loss, 2)}"
+                )
+            # Check for convergence
+            if check_for_convergence and i > min_num_iters:
+                if optimizer.converged():
+                    if show_progress_bar:
+                        print(f"\nConverged with loss: {np.round(mean_loss, 2)}")
+                    break
+        # Training finished:
+        self._trained_on = x
+
+        # Evaluate quality
+        if show_progress_bar:
```

Comment:
how about adding an additional parameter `quality_control`? It's a bit confusing that quality control is run if `show_progress_bar==True`

Reply:
Maybe even another method, as currently the only way to do quality controll is to run train.

Reply:
great idea, yes!

Reply:
I am also not sure which to use as default. As I experienced the PSIS does more distingish "very bad" from "ok", whereas the direct proportionality check does more check for "ok" and "good".

Reply:
Not sure which one to use as default. But please add exactly these sentences to the docstring ;)